File: C++/0191-number-of-1-bits.cpp
// Time:  O(1), atmost 32 bits
// Space: O(1)
class Solution {
public:
    int hammingWeight(uint32_t n) {
        int c;
        for (c = 0; n; n >>= 1) {
            c += n & 1;
        }
        return c;
    }
};



// Brian Kernigan's bit counting
class Solution {
public:
    int hammingWeight(uint32_t n) {
        int c;
        // only as many iterations as there are bits set
        for (c = 0; n; ++c) {
            // zeroing out the least significant nonzero bit
            n = n & (n - 1);
        }
        return c;
    }
};
/*
Notes:
Using built-in function: return __builtin_popcount(n);
*/

File: Python/0191-number-of-1-bits.py
# Time:  O(logn) = O(32)
# Space: O(1)
class Solution:
    def hammingWeight(self, n: int) -> int:
        count = 0
        while n:
            n &= n-1
            count += 1
        return count

"""
Notes: Here we are using bit operation to cancel a 1 in each round.
Think of a number in binary n = XXXXXX1000, n - 1 is XXXXXX0111.
n & (n - 1) will be XXXXXX0000 which is just cancel the last 1

Using built in function : return bin(n).count('1')
"""
